### PR TITLE
Ws 045 bulk upload

### PR DESCRIPTION
**Pull request by Eyrim**

### CODE REVIEW

File: src/main/java/com/whalespottingjava/controllers/ApiController.java [hunk 1]
```diff
@@ -3,10 +3,11 @@
 import com.whalespottingjava.models.database.Sighting;
 import com.whalespottingjava.services.SightingService;
 import io.swagger.v3.oas.annotations.Operation;
+import jakarta.validation.Valid;
+import org.aspectj.lang.annotation.AdviceName;
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/controllers/ApiController.java]: We're not using this import, so you can delete it

File: src/main/java/com/whalespottingjava/controllers/ApiController.java [hunk 2]
```diff
@@ -25,4 +26,16 @@ public ApiController(SightingService sightingService) {
     public List<Sighting> getAllSightings() {
         return this.sightingService.getAllApprovedSightings();
     }
+
+    @Operation(description = "Facilitates mass uploading of sightings to be approved")
+    @PostMapping("/api/bulk")
+    @ResponseStatus(HttpStatus.CREATED)
+//    @io.swagger.v3.oas.annotations.parameters.RequestBody
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/controllers/ApiController.java]: Do we need this comment for swagger annotations or can we delete it?

File: src/main/java/com/whalespottingjava/controllers/ApiController.java [hunk 3]
```diff
@@ -25,4 +26,16 @@ public ApiController(SightingService sightingService) {
...
+    public void postBulkSightings(
+            @RequestBody
+            @Valid
+            List<Sighting> sightings // TODO: 12/10/2023 JSON Hijacking?
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/controllers/ApiController.java]: Do you have a ticket to track this TODO? Or will you do it as part of this ticket?

File: src/main/java/com/whalespottingjava/models/database/Sighting.java [hunk 4]
```diff
@@ -1,8 +1,12 @@
 package com.whalespottingjava.models.database;
 
-import jakarta.persistence.*;
 import java.sql.Date;
 
+import io.swagger.v3.oas.annotations.Hidden;
+import jakarta.persistence.*;
+import jakarta.validation.Valid;
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/models/database/Sighting.java]: unused import

File: src/main/java/com/whalespottingjava/models/database/Sighting.java [hunk 5]
```diff
@@ -1,8 +1,12 @@
 package com.whalespottingjava.models.database;
 
-import jakarta.persistence.*;
 import java.sql.Date;
 
+import io.swagger.v3.oas.annotations.Hidden;
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/models/database/Sighting.java:3]: Can we remove the newline above this line?

File: src/main/java/com/whalespottingjava/controllers/ApiControllerAdvice.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+@RestControllerAdvice
+public class ApiControllerAdvice {
+    @ExceptionHandler({ DataIntegrityViolationException.class, ConstraintViolationException.class })
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public String dataIntegrityViolationHandler(Exception ex) {
+        return ex.getLocalizedMessage(); // TODO: 16/10/2023 This reveals implementation details on the database, error messages should be overridden
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/controllers/ApiControllerAdvice.java]: Ditto about this TODO (replace date with ticket number that covers implementing this behaviour)

File: src/main/java/com/whalespottingjava/services/SightingService.java [hunk 7]
```diff
@@ -7,6 +7,8 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.sql.Date;
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/services/SightingService.java]: unused import, also can we remove new line above?

File: src/main/java/com/whalespottingjava/services/SightingService.java [hunk 8]
```diff
@@ -7,6 +7,8 @@
...
+import java.time.LocalDate;
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/services/SightingService.java]: unused import

File: src/main/java/com/whalespottingjava/services/SightingService.java [hunk 9]
```diff
@@ -32,6 +34,12 @@ public List<Sighting> getAllApprovedSightings() { // TODO: 12/10/2023 Returns ra
         return this.sightingRepository.findByApprovedTrue();
     }
 
+    public void addBulkSightings(List<Sighting> sightings) {
+        // Validate sightings (Use jakarta?)
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/services/SightingService.java]: is this validation now done? I.e. can we remove the comment?

File: src/main/java/com/whalespottingjava/services/SightingService.java [hunk 10]
```diff
@@ -32,6 +34,12 @@ public List<Sighting> getAllApprovedSightings() { // TODO: 12/10/2023 Returns ra
...
+        // Add to DB
```
Comment (Wiesje) [inline src/main/java/com/whalespottingjava/services/SightingService.java]: This is now done so comment can be removed

**Review by Wiesje — COMMENTED** [review]
Nice work! Just some minor/styling comments.

This is exactly the kind of work that would give me peace of mind if we had a unit test that calls the endpoint with a request body, so we can see if the validation behaves as expected. (rather than checking if it has added data to the database, thought, we'd check just if it calls saveAll). 

Could you add markups for this PR and merge it, then create a new commit on this branch (or a new branch if you prefer) looking to get a couple unit tests written (at least 1 happy and 1 unhappy path)?

File: src/main/java/com/whalespottingjava/controllers/ApiController.java (hunk 3 above)
Comment (Eyrim) [reply src/main/java/com/whalespottingjava/controllers/ApiController.java]: Ticket added

File: src/main/java/com/whalespottingjava/controllers/ApiControllerAdvice.java (hunk 6 above)
Comment (Eyrim) [reply src/main/java/com/whalespottingjava/controllers/ApiControllerAdvice.java]: Ticket added

File: src/main/java/com/whalespottingjava/services/SightingService.java (hunk 9 above)
Comment (Eyrim) [reply src/main/java/com/whalespottingjava/services/SightingService.java]: The validation is not done yet, there is a ticket to track the issue, it's my next job after this merge. [Ticket.](https://github.com/users/charlottemothersole/projects/1/views/1?pane=issue&itemId=41634091)

**Review by Annie1790 — APPROVED** [review]

**Pull request merged**